File: src/DFRobot_EnvironmentalSensor.cpp
/*!
 * @file  DFRobot_EnvironmentalSensor.cpp
 * @brief Basic structure of DFRobot_EnvironmentalSensor class, the implementation of basic method
 *
 * @copyright	Copyright (c) 2021 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @license   The MIT License (MIT)
 * @author    [TangJie](jie.tang@dfrobot.com)
 * @version   V1.0
 * @date      2021-08-31
 * @url       https://github.com/DFRobot/DFRobot_EnvironmentalSensor
 */

#include "DFRobot_EnvironmentalSensor.h"

#ifdef __AVR__
typedef uint16_t    platformBitWidth_t;
#else
typedef uint32_t    platformBitWidth_t;
#endif

DFRobot_EnvironmentalSensor::DFRobot_EnvironmentalSensor(uint8_t addr, TwoWire *pWire)
{
  _pWire = pWire;
  _addr = addr;
};

DFRobot_EnvironmentalSensor::DFRobot_EnvironmentalSensor(uint8_t addr, Stream *s):DFRobot_RTU(s)
{
  _s = s;
  _addr = addr;
}

int8_t DFRobot_EnvironmentalSensor::begin(void)
{
  delay(500);
  setTimeoutTimeMs(200);
  if(_addr > 0xF7){
    DBG("Invaild Device addr.");
  }
  if(_addr != 0){
    if(!detectDeviceAddress(_addr)){
      DBG("Device addr Error.");
      return -1;
    }
  }else{
    return -1;
  }
  return 0;
}

uint16_t DFRobot_EnvironmentalSensor::getDevicePID(void)
{
  uint8_t buf[2];
  if(_pWire){
    readReg(REG_PID, buf, 2);
  }else{
    readReg(REG_PID, buf, 2);
  }
  return (buf[0] << 8 | buf[1]);

}

uint16_t DFRobot_EnvironmentalSensor::getDeviceVID(void)
{
  uint8_t buf[2];
  if(_pWire){
    readReg(REG_VID, buf, 2);
  }else{
    readReg(REG_VID, buf, 2);
  }
  return (buf[0] << 8 | buf[1]);

}

float DFRobot_EnvironmentalSensor::getTemperature(uint8_t unist)
{
  uint8_t buffer[2];
  uint16_t data;
  float temp;
  readReg(REG_TEMP,buffer,2);
  data = buffer[0] << 8 | buffer[1];
  temp = (-45) +((data * 175.00) / 1024.00 / 64.00);
  if(unist == TEMP_F){
    temp = temp * 1.8 + 32 ;
  }
  return temp;
}

float DFRobot_EnvironmentalSensor::getHumidity(void)
{
  uint8_t buffer[2];
  float humidity;
  uint16_t data;
  readReg(REG_HUMIDITY,buffer, 2);
  data = ((buffer[0] << 8 | buffer[1]));
  humidity = (float)data * 100 / 65536;
  return humidity;
}

float DFRobot_EnvironmentalSensor::mapfloat(float x, float in_min, float in_max, float out_min, float out_max)
{
      return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
}

float DFRobot_EnvironmentalSensor::getUltravioletIntensity(void)
{
  uint16_t uvLevel;
  uint16_t version = 0;
  uint8_t buffer[2];
  float ultraviolet;
  readReg(0x05,buffer,2);
  version = buffer[0] << 8 | buffer[1];
  DBG(buffer[0] << 8 | buffer[1]);
  if(version == 0x1001){
    readReg(REG_ULTRAVIOLET_INTENSITY,buffer,2);
    uvLevel = buffer[0] << 8 | buffer[1];
    DBG("A");
    DBG(uvLevel);
    ultraviolet = (float)uvLevel/1800.0;
  }else{
    readReg(REG_ULTRAVIOLET_INTENSITY,buffer,2);
    uvLevel = buffer[0] << 8 | buffer[1];
    float outputVoltage = 3.0 * uvLevel/1024;
    if(outputVoltage <= 0.99)
      outputVoltage = 0.99;
    else if(outputVoltage >= 2.99)
      outputVoltage = 2.99;
    ultraviolet = mapfloat(outputVoltage, 0.99, 2.9, 0.0, 15.0);
  }
  
  return ultraviolet;
}

float DFRobot_EnvironmentalSensor::getLuminousIntensity(void)
{
  uint16_t data;
  uint8_t buffer[2];
  readReg(REG_LUMINOUS_INTENSITY, buffer, 2);

  data = buffer[0] << 8 | buffer[1];
  float luminous = data;
	luminous = luminous * (1.0023f + luminous * (8.1488e-5f + luminous * (-9.3924e-9f + luminous * 6.0135e-13f)));
  return luminous;
}

uint16_t DFRobot_EnvironmentalSensor::getAtmospherePressure(uint8_t units)
{
  uint16_t atmosphere;
  uint8_t buffer[2];
  readReg(REG_ATMOSPHERIC_PRESSURE, buffer, 2); 
  atmosphere = buffer[0] << 8 | buffer[1];
  if(units == KPA){
    atmosphere /=10;
  }
  return atmosphere;
}

float DFRobot_EnvironmentalSensor::getElevation(void)
{
  uint16_t atmosphere;
  uint8_t buffer[2];
  readReg(REG_ATMOSPHERIC_PRESSURE, buffer, 2);
  atmosphere = buffer[0] << 8 | buffer[1];
 float elevation = 44330 * (1.0 - pow(atmosphere / 1015.0f, 0.1903));
  return elevation;
}

uint8_t DFRobot_EnvironmentalSensor::readReg(uint16_t reg, void *pBuf, uint8_t size)
{
  uint8_t* _pBuf = (uint8_t*)pBuf;
  
    if(pBuf == NULL){
      DBG("data error");
      return 0;
    }
  if(_pWire){
    uint8_t _reg = reg * 2;
    _pWire->begin();
    _pWire->beginTransmission(_addr);
    _pWire->write(_reg);
    _pWire->endTransmission();
    _pWire->requestFrom(_addr, size);
    for(uint8_t i = 0; i < size; i++)
      _pBuf[i] = _pWire->read();
      
    return size;
  }else{
    return readInputRegister(_addr, reg, _pBuf, size);
  }
}

bool  DFRobot_EnvironmentalSensor::detectDeviceAddress(uint8_t addr)
{
   if(_pWire){
   uint8_t buf[2];
   readReg(REG_DEVICE_ADDR, buf, 2);
   if(addr == ((buf[0] << 8 | buf[1]) & 0xFF))
    return true;
    
  }else{
    uint16_t ret = readInputRegister(addr, REG_DEVICE_ADDR);
    DBG(ret);
    if((ret & 0xFF) == addr)
      return true;
  }
  return false;
}

File: python/raspberry/DFRobot_Environmental_Sensor.py
# -*- coding: utf-8 -*
'''!
  @file        dfrobot_environmental_sensor.py
  @brief       DFRobot_EnvironmentalSensor library is developed for DFRobot environmental sensor.
  @copyright   Copyright (c) 2021 DFRobot Co.Ltd (http://www.dfrobot.com)
  @license     The MIT License (MIT)
  @author      TangJie(jie.tang@dfrobot.com)
  @version     V1.0
  @date        2021-08-31
  @url         https://github.com/DFRobot/DFRobot_EnvironmentalSensor
'''

import serial
import time
import smbus
import os
import math
import RPi.GPIO as GPIO
import math

import modbus_tk
import modbus_tk.defines as cst
from modbus_tk import modbus_rtu

I2C_MODE                  = 0x01
UART_MODE                 = 0x02
DEV_ADDRESS               = 0x22

DEVICE_VID                = 0x3343
DEVICE_ADDRESS            = 0x22

HPA                       = 0x01
KPA                       = 0X02
TEMP_C                    = 0X03
TEMP_F                    = 0X04

class DFRobot_Environmental_Sensor():
  '''!
    @brief Define DFRobot_Environmental_Sensor basic class
    @details For driving the environmental sensor
  '''
  def __init__(self ,bus ,baud = 9600):
    if bus != 0:
      self.i2cbus = smbus.SMBus(bus)
      self._uart_i2c = I2C_MODE
      
    else:
      self.master = modbus_rtu.RtuMaster(serial.Serial(port="/dev/ttyAMA0",baudrate=baud, bytesize=8, parity='N', stopbits=1))
      self.master.set_timeout(1.0)
      self._uart_i2c = UART_MODE
  
  def _detect_device_address(self):
    '''!
      @brief Get sensor address
      @return  Return sensor address
    '''
    rbuf = self._read_reg(0x04,2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    return data

  
  def begin(self):
    '''!
      @brief Init SEN0500/SEN0501 sensor
      @return Return init status
      @retval 0  Succeed
      @retval -1 Failed
    '''
    if self._detect_device_address() != DEV_ADDRESS:
      return False
    return True

  
  def get_temperature(self,unist):
    '''!
      @brief Get SEN0500/SEN0501 temperature data
      @param units Temperature data unit select
      @n     TEMP_C ℃
      @n     TEMP_F ℉ 
      @return Return the obtained temperature data
    '''
    rbuf = self._read_reg(0x14, 2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    temp = (-45) +((data * 175.00) / 1024.00 / 64.00)
    if(unist == TEMP_F):
      temp = temp * 1.8 + 32 
    return round(temp,2)
    
  
  def get_humidity(self):
    '''!
      @brief Get SEN0500/SEN0501 humidity data 
      @return Return the obtained humidity data
    '''
    rbuf = self._read_reg(0x16, 2)
    if self._uart_i2c == I2C_MODE:
      humidity = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      humidity = rbuf[0]
    humidity = (humidity / 1024) * 100 / 64
    return humidity
  
 
  def get_ultraviolet_intensity(self):
    '''!
      @brief Get SEN0500/SEN0501 UV intensity index data 
      @return Return the obtained UV intensity index data
    '''
    version = self.self._read_reg(0x05, 2)
    if (version[0] << 8 | version[1]) == 0x1001:
      rbuf = self._read_reg(0x10, 2)
      data = rbuf[0] << 8 | rbuf[1]
      ultraviolet = data / 1800
    else:
      rbuf = self._read_reg(0x10, 2)
      if self._uart_i2c == I2C_MODE:
        data = rbuf[0] << 8 | rbuf[1]
      elif self._uart_i2c == UART_MODE:
        data = rbuf[0]
      outputVoltage = 3.0 * data/1024
      ultraviolet = (outputVoltage - 0.99) * (15.0 - 0.0) / (2.9 - 0.99) + 0.0 
    return round(ultraviolet,2)
      
  
  def get_luminousintensity(self):
    '''!
      @brief Get SEN0500/SEN0501 luminous intensity data 
      @return Return the obtained luminous intensity data
    '''
    rbuf = self._read_reg(0x12 ,2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    luminous = data * (1.0023 + data * (8.1488e-5 + data * (-9.3924e-9 + data * 6.0135e-13)))
    return round(luminous,2)

  
  def get_atmosphere_pressure(self, units):
    '''!
      @brief Get SEN0500/SEN0501 atmosphere pressure data 
      @param units Atmosphere pressure data unit select
      @n            HPA Hectopascal
      @n            KPA Kilopascal
      @return Return the obtained atmosphere pressure data
    '''
    rbuf = self._read_reg(0x18, 2)
    if self._uart_i2c == I2C_MODE:
      atmosphere = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      atmosphere = rbuf[0]
    if units == KPA:
      atmosphere /= 10
    return atmosphere

  
  def get_elevation(self):
    '''!
      @brief Get SEN0500/SEN0501 elevation data 
      @return Return the obtained elevation data
    '''
    rbuf = self._read_reg(0x18, 2)
    if self._uart_i2c == I2C_MODE:
      elevation = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      elevation = rbuf[0]
    elevation = 44330 * (1.0 - pow(elevation / 1015.0, 0.1903));
    return round(elevation,2)

  
        

class DFRobot_Environmental_Sensor_I2C(DFRobot_Environmental_Sensor):
  '''!
    @brief An example of an i2c interface module
  '''
  def __init__(self ,bus ,addr):
    self._addr = addr
    DFRobot_Environmental_Sensor.__init__(self,bus,0)   
    
  
  def _read_reg(self, reg_addr ,length):
    '''!
      @brief read the data from the register
      @param reg register address
      @param value read data
    '''
    try:
      rslt = self.i2cbus.read_i2c_block_data(self._addr ,reg_addr , length)
    except:
      rslt = -1
    return rslt    


class DFRobot_Environmental_Sensor_UART(DFRobot_Environmental_Sensor):
  '''!
    @brief An example of an UART interface module
  '''
  def __init__(self ,baud, addr):
    self._baud = baud
    self._addr = addr
    try:
      DFRobot_Environmental_Sensor.__init__(self,0,self._baud)
    except:
      print ("plese get root!")
   
  
  def _read_reg(self, reg_addr ,length):
    '''!
      @brief Read data from the sensor
    '''
    return list(self.master.execute(self._addr, cst.READ_INPUT_REGISTERS, reg_addr/2, length/2))
    
    
